grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -0,0 +1,323 @@
+    synchronized (lock) {
+      if (closed) {
+        return;
+      }
+      closed = true;
+      if (lbRequestWriter != null) {
+        lbRequestWriter.onCompleted();
+      }
+      executor = SharedResourceHolder.release(GrpcUtil.SHARED_CHANNEL_EXECUTOR, executor);
+      deadlineCancellationExecutor = SharedResourceHolder.release(
+          GrpcUtil.TIMER_SERVICE, deadlineCancellationExecutor);
+    }
+  }
+
+  @Override
+  public void transportShutdown(
+      EquivalentAddressGroup addressGroup, ClientTransport transport, Status status) {
+    handleError(status.augmentDescription("Transport to LB server closed"));
+    synchronized (lock) {
+      if (transport == lbTransport) {
+        connectToLb();
+      }
+    }
+  }
+
+  private void handleError(Status error) {
+    FulfillmentBatch<ClientTransport> pendingPicksFulfillmentBatch;
+    StatusException statusException = error.asException();
+    synchronized (lock) {
+      lastError = statusException;
+      pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
+    }
+    pendingPicksFulfillmentBatch.fail(statusException);
+  }
+
+  private class LbResponseObserver implements StreamObserver<LoadBalanceResponse> {
+    @Override public void onNext(LoadBalanceResponse response) {
+      logger.info("Got a LB response: " + response);
+      FulfillmentBatch<ClientTransport> pendingPicksFulfillmentBatch;
+      final RoundRobinServerList newRoundRobinServerList;
+      HashMap<SocketAddress, ResolvedServerInfo> newServerMap =
+          new HashMap<SocketAddress, ResolvedServerInfo>();
+      synchronized (lock) {
+        InitialLoadBalanceResponse initialResponse = response.getInitialResponse();
+        // TODO(zhangkun83): make use of initialResponse
+        RoundRobinServerList.Builder listBuilder = new RoundRobinServerList.Builder(tm);
+        ServerList serverList = response.getServerList();
+        // TODO(zhangkun83): honor expiration_interval
+        for (Server server : serverList.getServersList()) {
+          if (server.getDropRequest()) {
+            listBuilder.add(null);
+          } else {
+            InetSocketAddress address = new InetSocketAddress(
+                server.getIpAddress(), server.getPort());
+            listBuilder.add(address);
+            // TODO(zhangkun83): fill the LB token to the attributes, and insert it to the
+            // application RPCs.
+            if (!newServerMap.containsKey(address)) {
+              newServerMap.put(address, new ResolvedServerInfo(address, Attributes.EMPTY));
+            }
+          }
+        }
+        newRoundRobinServerList = listBuilder.build();
+        if (newRoundRobinServerList.size() == 0) {
+          return;
+        }
+        roundRobinServerList = newRoundRobinServerList;
+        servers = newServerMap;
+        pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
+      }
+      tm.updateRetainedTransports(newServerMap.keySet());
```

Comment:
Is there any concern about interactions with this call to retain LB'ed transports and the transport used to obtain LB information? It seems like it _could_ be fine, but I've not thought about it very long.